src/JZSolver.cpp

```diff
@@ -0,0 +1,106 @@
+#include "JZSolver.hpp"
+#include <numeric>
+#include <iostream>
+
+bool JZSolver::solve(std::shared_ptr<Board> b)
+{
+    verifyBoardClues(b);
+    findAndFillCompleteLines(b);
+
+    while(not b->isSolved())
+    {
+        stepCounter_++;
+        // methods to solve puzzle
+        // if method changes something at board then start loop from beginning
+        // if all methods haven't changed anything then break loop and admit failure
+        break;
+    }
+    return b->isSolved();
+}
+
+unsigned int JZSolver::getStepCounter()
+{
+    return stepCounter_;
+}
+
+void JZSolver::verifyBoardClues(std::shared_ptr<Board> b)
```

Comment:
This function is universal to every algorithm. It should be called before any other algorithm is used. And I think it should be in a Board class. Board constructor can prevent creating a board which has invalid clues. Thanks to that you will not need to check it in your algorithms.

Reply:
That was also my original thought on this topic, but then I realised that adding it to Board constructor would destroy all of our testing cases.

Reply:
That's good. So you need to change tests to adjust them to new behaviour.